File: connectors/mongo/sync/src/main/java/com/coditory/sherlock/mongo/MongoCollectionInitializer.java
package com.coditory.sherlock.mongo;

import com.mongodb.client.MongoCollection;
import org.bson.Document;

import java.util.concurrent.atomic.AtomicBoolean;

import static com.coditory.sherlock.Preconditions.expectNonNull;
import static com.coditory.sherlock.mongo.MongoDistributedLock.INDEX;
import static com.coditory.sherlock.mongo.MongoDistributedLock.INDEX_OPTIONS;

class MongoCollectionInitializer {
    private final MongoCollection<Document> collection;
    private final AtomicBoolean indexesCreated = new AtomicBoolean(false);

    MongoCollectionInitializer(MongoCollection<Document> collection) {
        expectNonNull(collection, "collection");
        this.collection = collection;
    }

    MongoCollection<Document> getInitializedCollection() {
        boolean shouldCreateIndexes = indexesCreated.compareAndSet(false, true);
        if (!shouldCreateIndexes) {
            return collection;
        }
        collection.createIndex(INDEX, INDEX_OPTIONS);
        return collection;
    }
}

File: connectors/mongo/coroutines/src/main/kotlin/com/coditory/sherlock/mongo/coroutines/MongoCollectionInitializer.kt
package com.coditory.sherlock.mongo.coroutines

import com.coditory.sherlock.mongo.MongoDistributedLock.INDEX
import com.coditory.sherlock.mongo.MongoDistributedLock.INDEX_OPTIONS
import com.mongodb.kotlin.client.coroutine.MongoCollection
import org.bson.Document
import java.util.concurrent.atomic.AtomicBoolean

internal class MongoCollectionInitializer(
    private val collection: MongoCollection<Document>,
) {
    private val indexesCreated = AtomicBoolean(false)

    suspend fun getInitializedCollection(): MongoCollection<Document> {
        val shouldCreateIndexes = indexesCreated.compareAndSet(false, true)
        if (!shouldCreateIndexes) {
            return collection
        }
        collection.createIndex(INDEX, INDEX_OPTIONS)
        return collection
    }
}
